File: src/mod/class/hs-modules/hs-heater/temp-helper-files/heater_compare.js
const xlsx = require('xlsx');
const path = require('path');
const fs = require('fs');

const workbookPath = path.resolve(__dirname, '../../../../../../docs/SynergismOfficial-master/Re-Octptimisation (Beta 1.2.5E)/Re-Octptimisation (Beta 1.2.5E).xlsx');
const sheetName = 'True Ambrosia Heater';
const labelToField = {
    'Asc. Speed (C)': 'ascSpeed',
    'Ascension Speed (H)': 'ascSpeed2',
    'SI Non-Amb RC': 'sirc',
    'Bonus SI': 'bonussi',
    'Bonus IA (total)': 'totalbonusia',
    'Bonus IA (from talisman)': 'talismanbonusia',
    'Base Talisman Power': 'baseTalismanPower',
    'Rune Exp (log) (>50)': 'runeexp',
    'Blueberries': 'blueberries',
    'Base Obtainium': 'baseobt',
    'Base Offering': 'baseoff',
    'Voucher': 'voucher',
};

if (!fs.existsSync(workbookPath)) {
    console.error(`Workbook not found: ${workbookPath}`);
    process.exit(1);
}

const wb = xlsx.readFile(workbookPath, { cellStyles: false });
if (!wb.SheetNames.includes(sheetName)) {
    console.error('Sheet not found. Available sheets:', wb.SheetNames.join(', '));
    process.exit(1);
}

const sheet = wb.Sheets[sheetName];
const range = xlsx.utils.decode_range(sheet['!ref']);
const rows = [];
for (let r = range.s.r; r <= range.e.r; ++r) {
    const row = [];
    for (let c = range.s.c; c <= range.e.c; ++c) {
        const addr = xlsx.utils.encode_cell({ r, c });
        const cell = sheet[addr];
        if (!cell) {
            row.push(null);
        } else if (cell.f) {
            row.push(`f:${cell.f}`);
        } else {
            row.push(cell.v);
        }
    }
    rows.push(row);
}

const matches = [];
for (let r = 0; r < rows.length; ++r) {
    const row = rows[r];
    for (let c = 0; c < row.length; ++c) {
        const cell = row[c];
        if (typeof cell === 'string' && Object.keys(labelToField).includes(cell.trim())) {
            const label = cell.trim();
            const field = labelToField[label];
            const tail = row.slice(c + 1, c + 12);
            const firstNumeric = tail.find((v) => typeof v === 'number');
            matches.push({ row: r + 1, col: c + 1, label, field, firstNumeric, tail });
            break;
        }
    }
}

console.log('Spreadsheet comparison rows for target labels:');
for (const match of matches) {
    console.log('---');
    console.log(`Row ${match.row}, Col ${match.col}, Label: ${match.label}, Field: ${match.field}`);
    console.log(`First numeric after label: ${match.firstNumeric}`);
    console.log('Tail values:', JSON.stringify(match.tail.slice(0, 10)));
}

const samplePath = process.argv[2];
if (samplePath) {
    const resolvedSamplePath = path.resolve(process.cwd(), samplePath);
    if (!fs.existsSync(resolvedSamplePath)) {
        console.error('Sample export path not found:', resolvedSamplePath);
        process.exit(1);
    }
    const sampleRaw = fs.readFileSync(resolvedSamplePath, 'utf-8');
    const sample = JSON.parse(sampleRaw);
    const exportHsData = sample.hs_data;

    console.log('\nComparing sample export values against spreadsheet values:');
    for (const match of matches) {
        const actual = exportHsData ? exportHsData[match.field] : undefined;
        console.log(`${match.field}: sheet=${match.firstNumeric} export=${actual}`);
    }
}

if (matches.length === 0) {
    console.error('No matching row labels found in the spreadsheet.');
    process.exit(1);
}

File: src/mod/class/hs-modules/hs-heater/temp-helper-files/heater_compare.py
from pathlib import Path
import json
import openpyxl
import argparse

WORKBOOK_RELATIVE = Path("../../../../../../docs/SynergismOfficial-master/Re-Octptimisation (Beta 1.2.5E)/Re-Octptimisation (Beta 1.2.5E).xlsx")
SHEET_NAME = "True Ambrosia Heater"
LABEL_TO_FIELD = {
    'Asc. Speed (C)': 'ascSpeed',
    'Ascension Speed (H)': 'ascSpeed2',
    'SI Non-Amb RC': 'sirc',
    'Bonus SI': 'bonussi',
    'Bonus IA (total)': 'totalbonusia',
    'Bonus IA (from talisman)': 'talismanbonusia',
    'Base Talisman Power': 'baseTalismanPower',
    'Rune Exp (log) (>50)': 'runeexp',
    'Blueberries': 'blueberries',
    'Base Obtainium': 'baseobt',
    'Base Offering': 'baseoff',
    'Voucher': 'voucher',
}


def read_sheet_values(workbook_path: Path):
    if not workbook_path.exists():
        raise FileNotFoundError(f"Workbook not found: {workbook_path}")

    wb = openpyxl.load_workbook(workbook_path, data_only=True, read_only=True)
    if SHEET_NAME not in wb.sheetnames:
        raise ValueError(f"Sheet '{SHEET_NAME}' not found. Available sheets: {wb.sheetnames}")

    sheet = wb[SHEET_NAME]
    found = {}
    for row in sheet.iter_rows(values_only=True):
        label = row[0]
        if label is None:
            continue
        label = str(label).strip()
        if label in LABEL_TO_FIELD:
            field = LABEL_TO_FIELD[label]
            found[field] = {
                'label': label,
                'value': row[1] if len(row) > 1 else None,
                'ref': row[4] if len(row) > 4 else None,
            }
    return found


def compare_export(sample_path: Path, sheet_values: dict):
    if not sample_path.exists():
        raise FileNotFoundError(f"Sample export not found: {sample_path}")
    sample = json.loads(sample_path.read_text(encoding='utf-8'))
    if 'hs_data' not in sample:
        raise ValueError('Sample JSON missing hs_data root object')
    export = sample['hs_data']

    results = []
    for field, metadata in sheet_values.items():
        actual = export.get(field)
        results.append({
            'field': field,
            'label': metadata['label'],
            'sheet_value': metadata['value'],
            'export_value': actual,
            'match': actual == metadata['value'],
        })
    return results


def main():
    parser = argparse.ArgumentParser(description='Compare Ambrosia Heater export fields against the True Ambrosia Heater spreadsheet.')
    parser.add_argument('--sample', type=Path, help='Path to a JSON heater export sample file')
    args = parser.parse_args()

    workbook_path = (Path(__file__).resolve().parent / WORKBOOK_RELATIVE).resolve()
    sheet_values = read_sheet_values(workbook_path)

    print('Spreadsheet values from', workbook_path)
    for field, metadata in sheet_values.items():
        print(f"{field}: label={metadata['label']} value={metadata['value']} ref={metadata['ref']}")

    if args.sample:
        print('\nComparing to sample export', args.sample)
        results = compare_export(args.sample, sheet_values)
        for result in results:
            status = 'MATCH' if result['match'] else 'MISMATCH'
            print(f"{status}: {result['field']} sheet={result['sheet_value']} export={result['export_value']}")


if __name__ == '__main__':
    main()
